VideoClassification/model.py

```diff
@@ -0,0 +1,267 @@
+	2D tensor with shape: `(batch_size, feature_dim)`.
+	"""
+
+	def __init__(self, input_shape, **kwargs):
+		super(ContextGating, self).__init__(**kwargs)
+		feature_dim = input_shape[-1]
+		if not isinstance(feature_dim, int):
+			feature_dim = feature_dim.value
+		self.fc = tf.keras.layers.Dense(
+			units=feature_dim,
+			activation=tf.nn.sigmoid,
+			kernel_regularizer=tf.keras.regularizers.l2(1e-5),
+		)
+
+	def call(self, model_input):
+		"""Apply the ContextGating module to the given input.
+
+		Args:
+			model_input: A tensor with shape [batch_size, feature_dim].
+		Returns:
+			A tensor with shape [batch_size, feature_dim].
+		Raises:
+			ValueError: If the `feature_dim` of model_input is not defined.
+		"""
+		model_input.shape.assert_has_rank(2)
+		feature_dim = model_input.shape.as_list()[-1]
+		if feature_dim is None:
+			raise ValueError("Last dimension must be defined.")
+		
+		context_gate = self.fc(model_input)
+
+		output = tf.math.multiply(context_gate, model_input)
+
+		return output
+
+	def compute_output_shape(self, input_shape):
+		return input_shape
+
+	def get_config(self):
+		base_config = super().get_config()
+		return dict(list(base_config.items()))
+
+class MOELogistic(tf.keras.layers.Layer):
+	"""Implements a Mixture of Logistic Experts classifier.
+
+	Input shape:
+		2D tensor with shape: `(batch_size, feature_dim)`.
+	Output shape:
+		2D tensor with shape: `(batch_size, num_classes)`.
+	"""
+
+	def __init__(self, input_shape, num_classes, num_mixtures, **kwargs):
+		super(MOELogistic, self).__init__(**kwargs)
+		self.num_classes = num_classes
+		self.num_mixtures = num_mixtures
+
+		self.gate_fc = tf.keras.layers.Dense(
+			units=num_classes*(num_mixtures+1),
+			kernel_regularizer=tf.keras.regularizers.l2(1e-6),
+		)
+
+		self.expert_fc = tf.keras.layers.Dense(
+			units=num_classes*num_mixtures,
+			kernel_regularizer=tf.keras.regularizers.l2(1e-6),
+		)
+
+	def call(self, input):
+		"""Apply the MoE algorithm to the given input.
+
+		Args:
+			input: A tensor with shape [batch_size, feature_dim].
+		Returns:
+			A tensor with shape [batch_size, num_classes].
+		Raises:
+			ValueError: If the `feature_dim` of input is not defined.
+		"""
+		gate_activations = self.gate_fc(input)
+		expert_activations = self.expert_fc(input)
+
+		#Calculate the distribution across mixtures
+		gate_dist = tf.nn.softmax(tf.reshape(gate_activations, [-1, self.num_mixtures+1]))
+		expert_dist = tf.nn.sigmoid(tf.reshape(expert_activations, [-1, self.num_mixtures]))
+
+		probs = tf.reduce_sum(tf.math.multiply(gate_dist[:,:self.num_mixtures], expert_dist),1)
+		probs = tf.reshape(probs, [-1, self.num_classes])
+
+		return probs
+
+	def compute_output_shape(self, input_shape):
+		return (input_shape[0], self.num_classes)
+
+	def get_config(self):
+		base_config = super().get_config()
+		config = base_config.update({'number of classes': self.num_classes, 'number of mixtures': self.num_mixtures})
+		return config
+
+class VideoClassifier:
```

Comment:
Not about the code: can you also train a version that doesn't use audio features? That may help us understand better whether the class specific components help visual parts more or both.

Reply:
I can pretty easily train this classifier. However, the analysis will need to wait until after I have completed the segment level classifier as that is where I use class specific features